Update checking proxies

sssmaximusss opened the pull request:
Я переписал поиск прокси. Все прокси закидываю в ExecutorService, запускаю поиск для каждого и параллельно дальше работаю. Если прокси - рабочий, сразу записываю его в файл.
Не настраивал синхронизацию!!! Поэтому все(около 1000 проксей) набрасываются на ЖЖ одновременно. Но у каждого прокси свой запрос. Не знаю, будет ли это банится.
При таком сценарии все прокси проверяются за 10 секунд. Это число обусловлено таймаутом на чтение данных.

sssmaximusss committed:
rewrote code, which searches working proxies, using Callable and ExecutorService
release writing to file every working proxy

KorixD commented:
Замечание по общему виду (у Олега это не так актуально, у тебя актуальнее):
http://www.nurkiewicz.com/2013/05/java-8-definitive-guide-to.html 
0) Поиск работающих прокси с помощью CompetableFuture, это выглядит изящнее + с Future'ами действительно стоит научиться работать :) .
`

```
private void findWorkingProxy() {
    ExecutorService service = Executors.newCachedThreadPool();
    Iterator<String> iterator = rawAllUsers.iterator();
    String user = iterator.next();
    rawProxies.forEach(
            proxy -> CompletableFuture.supplyAsync(() -> checkProxy(proxy, user), service)
                    .thenAccept(this::writeProxyToFile));
}

private ProxyWrapper checkProxy(HttpHost proxy, String nick) {
    boolean response;
    try {
        logger.info("Check proxy: " + proxy.toString());
        response = new ProxyLoader().loadData(
                new HttpHost(proxy.getHostName(), proxy.getPort()), nick);
    } catch (InterruptedException | UnirestException | IOException e) {
        logger.error("Error checking proxy: " + proxy.toString() + " to find info about user: " + nick + ". " + e);
        return new ProxyWrapper(proxy, false);
    }

    return new ProxyWrapper(proxy, response);
}

private synchronized void writeProxyToFile(ProxyWrapper proxyWrapper) {
    HttpHost proxy = proxyWrapper.proxy;

    if (proxyWrapper.isWorking) {
        if (!workingProxies.contains(proxy)) {
            workingProxies.add(proxy);
            try (PrintWriter printWriter = new PrintWriter(
                    new FileWriter(WORKING_PROXIES_FILE_FULL_PATH, true))) {
                printWriter.println(proxy.getHostName() + ":" + proxy.getPort());
            } catch (IOException e) {
                logger.error("Error writing to file: " + WORKING_PROXIES_FILE_FULL_PATH + ". " + e);
            }
        }
    } else {
        logger.info("No access to LJ for proxy: " + proxy.toString());
        brokenProxies.add(proxy);
    }
}

private static class ProxyWrapper {
    HttpHost proxy;
    boolean isWorking;

    public ProxyWrapper(HttpHost proxy, boolean isWorking) {
        this.proxy = proxy;
        this.isWorking = isWorking;
    }
}
```

`

1) setRawUsers в initCrawler в таком виде как есть, кажется, не нужно. Он запускается 1 раз и добавляет того же startUser, который уже есть в ProxyFactory.

KorixD commented on src/main/java/com/interestscsc/crawler/proxy/ProxyFactory.java:
```diff
@@ -26,8 +27,8 @@
     private Set<HttpHost> rawProxies;
-    private List<HttpHost> workingProxies;
-    private List<HttpHost> brokenProxies;
+    private static List<HttpHost> workingProxies;
```
А зачем?

sssmaximusss replied:
чтобы этот список использовать во вложенном классе ProxyChecker :)

sssmaximusss committed:
added changes to checking proxies
used CompetableFuture
fixed NPE which handling when we get next iterator of rawUserSet

sssmaximusss committed:
added method to get proxy-list from the website

sssmaximusss commented:
внес изменения в соответствии с предыдущими замечаниями.
Добавил получения списка прокси с сайта http://webanetlabs.net/freeproxy/proxylist_at_15.03.2016.txt
Безусловным минусом является то, что код привязан именно к этому сайту. Будут другие сайты, я пойму, что нужно поменять. А пока собираю актуальный по дате список проксей оттуда: из ~1000 работают 100-150.

KorixD commented on src/test/java/com/interestscsc/crawler/test/FindWorkingProxyTest.java:
```diff
@@ -36,11 +34,16 @@ public void findWorkingProxy() throws SQLException {
-        logger.info("Count of all proxies: " + proxyFactory.getCountRawProxies());
-        logger.info("Count of working proxies: " + proxyFactory.getCountWorkingProxies());
+        int session = 20;
+        while (session-- > 0) {
```
но почему не for?
исправишь - мерджи.

и не нужно делать дополнительный коммит)

sssmaximusss replied:
исправил - мерджу)
